File: Medium/RemoveNodesFromLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head
        prev = None
        current = head
        while current:
            next_temp = current.next
            current.next = prev
            prev = current
            current = next_temp

        # New Reversed list
        reverse_head = prev
        current = prev
        local_max = current.val
        while current.next:
            if current.next.val < local_max:
                current.next = current.next.next
            else:
                current = current.next
                local_max = max(local_max, current.val)

        # ReReversed list
        prev = None
        current = reverse_head
        while current:
            next_temp = current.next
            current.next = prev
            prev = current
            current = next_temp
        return prev

File: Medium/RemoveNodesFromLinkedList.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func removeNodes(_ head: ListNode?) -> ListNode? {
        // Corner Case
        guard let head = head else { return nil }

        // Reverse the list
        var prev: ListNode? = nil
        var current = head

        while let next = current.next {
            current.next = prev
            prev = current
            current = next
        }
        current.next = prev

        // Reversed List: Discard unwanted
        var newHead = current
        var localMax = newHead.val
        var currentNode = newHead

        while let nextNode = currentNode.next {
            if nextNode.val < localMax {
                currentNode.next = nextNode.next
            } else {
                localMax = max(localMax, nextNode.val)
                currentNode = nextNode
            }
        }

        // Reverse the list and return
        prev = nil
        current = newHead
        while let next = current.next {
            current.next = prev
            prev = current
            current = next
        }
        current.next = prev
        return current
    }
}